### ISSUE TITLE
speed

### ISSUE THREAD

**Issue by DanGrayson** [UniMath/UniMath#625]
Here are the times of the slowest 12 files (on my Mac), accounting for 68% of the time to compile all the UniMath *.v files:
```
user time 321.70: UniMath/HomologicalAlgebra/TranslationFunctors
user time 257.96: UniMath/HomologicalAlgebra/MappingCone
user time 251.80: UniMath/HomologicalAlgebra/KAPreTriangulated
user time 209.78: UniMath/SubstitutionSystems/LiftingInitial_alt
user time 190.03: UniMath/HomologicalAlgebra/MappingCylinder
user time 189.73: UniMath/HomologicalAlgebra/KA
user time 171.84: UniMath/HomologicalAlgebra/CohomologyComplex
user time 163.66: UniMath/SubstitutionSystems/MonadsFromSubstitutionSystems
user time 162.91: UniMath/HomologicalAlgebra/KATriangulated
user time 138.55: UniMath/SubstitutionSystems/LiftingInitial
user time 131.31: UniMath/RealNumbers/Reals
user time 117.19: UniMath/CategoryTheory/LocalizingClass
```
They are more than twice as slow as the next one on the list.

Here are the word counts of the same files:
```
    9041 UniMath/HomologicalAlgebra/TranslationFunctors.v
   34139 UniMath/HomologicalAlgebra/MappingCone.v
    3789 UniMath/HomologicalAlgebra/KAPreTriangulated.v
    3229 UniMath/SubstitutionSystems/LiftingInitial_alt.v
    5428 UniMath/HomologicalAlgebra/MappingCylinder.v
    3133 UniMath/HomologicalAlgebra/KA.v
   12616 UniMath/HomologicalAlgebra/CohomologyComplex.v
    3100 UniMath/SubstitutionSystems/MonadsFromSubstitutionSystems.v
    1004 UniMath/HomologicalAlgebra/KATriangulated.v
    3678 UniMath/SubstitutionSystems/LiftingInitial.v
    7906 UniMath/RealNumbers/Reals.v
   12241 UniMath/CategoryTheory/LocalizingClass.v
```

Can we improve our proofs are definitions to improve the speed?  It would be good to postpone the time when UniMath fissions so we can continue to make global improvements easily.

**Comment by DanGrayson** [comment]
Here are the speeds in seconds per thousand words:
```
      35  UniMath/HomologicalAlgebra/TranslationFunctors.v
       7  UniMath/HomologicalAlgebra/MappingCone.v
      66  UniMath/HomologicalAlgebra/KAPreTriangulated.v
      64  UniMath/SubstitutionSystems/LiftingInitial_alt.v
      35  UniMath/HomologicalAlgebra/MappingCylinder.v
      60  UniMath/HomologicalAlgebra/KA.v
      13  UniMath/HomologicalAlgebra/CohomologyComplex.v
      52  UniMath/SubstitutionSystems/MonadsFromSubstitutionSystems.v
     162  UniMath/HomologicalAlgebra/KATriangulated.v
      37  UniMath/SubstitutionSystems/LiftingInitial.v
      16  UniMath/RealNumbers/Reals.v
       9  UniMath/CategoryTheory/LocalizingClass.v
```

**Comment by rmatthes** [comment]
Concerning UniMath/SubstitutionSystems/MonadsFromSubstitutionSystems.v and UniMath/SubstitutionSystems/LiftingInitial.v, quite some effort went into bringing the execution time down to that tolerable(?) level. Maybe there is still potential in UniMath/SubstitutionSystems/LiftingInitial_alt.v that is based on cocontinuity instead of Kan extensions.

**Comment by mortberg** [comment]
When working on CT I spent a lot of time making sure things were reasonably fast. Here come some heuristics for how I typically achieved this:

- Use `Qed` and `abstract` as much as possible. The rationale is that the less things Coq has to compare the faster it will be. Proofs tend to be represented by huge terms and if Coq has to compare these over and over it gets very slow. Sometimes there are terms where the definition is important which still has to be made opaque for some proof to be fast enough, I would then use Opaque/Transparent, but this was quite rare in the proofs I did (here is one example of this: https://github.com/UniMath/UniMath/blob/master/UniMath/CategoryTheory/Inductives/Lists.v#L118). 

- Be careful with `simpl`. The reason is that things unfold too much and Coq has a hard time comparing things. I would typically write my proofs with a lot `simpl` and then take another pass over the proof script and remove as many of these as possible. I also sometimes noticed serious speed-ups of `Qed`/`Defined` when I used `cbn` instead of `simpl` in some places. 

- Replace `rewrite` by a combination of `etrans` and `apply`. For some reason I noticed that `rewrite` sometimes construct terms that are very slow for `Qed`/`Defined` and if I used a combination of `etrans` and `apply` things would get much faster. I guess the reason is that `rewrite` is not smart enough to find the simplest proofs all the time. 


I would also be careful to optimize my proofs when I was writing them. If a tactic application or Qed/Defined was slow (taking more than a second) I would spend some time optimizing things. This way the inefficiencies are found earlier and one won't end up with a `Qed` that suddenly takes a few minutes.

**Comment by mortberg** [comment]
I actually spent a few hours trying to optimize `LiftingInitial` and `LiftingInitial_alt` using the heuristics above, but I didn't make much progress. I think for these files there could be a mathematical way to make things faster, they are both proving the same thing but from different hypotheses. Can we make a more general version of this result so that LiftingInitial and LiftingInitial_alt are instrances?

It is often the case that proofs get faster when working with abstract hypotheses (as these can't be unfolded) compared to working with something too concrete.

**Comment by benediktahrens** [comment]
On 02/27/2017 02:02 PM, Daniel R. Grayson wrote:
> 
> Can we improve our proofs are definitions to improve the speed?  It
> would be good to postpone the time when UniMath fissions so we can
> continue to make global improvements easily.

I agree that that is an important goal.

Do you have any suggestions for how to achieve that goal?

**Comment by DanGrayson** [comment]
No, I have no precise suggestions.  The purpose of this issue is to keep it in our minds, to persuade people to share their ideas here.  If I start working on optimizing others' code, I'll assign myself to this issue, and others may do the same -- we can multiple people working on it.

**Comment by DanGrayson** [comment]
Twice in a week I've seen travis run out of time (49 minutes) checking HomologicalAlgebra.  So something needs to be done sooner than I thought.

**Comment by tpannila** [comment]
I see that most of the problematic files are written by me. I'm sorry but I don't know how to fix
them. Since HomologicalAlgebra causes Travis build timeouts, either Travis should not 
build HomologicalAlgebra or someone should pay for a better travis plan 
https://travis-ci.com/plans .

**Comment by mortberg** [comment]
Or we should figure out why these files are so slow and make them faster. Take a look at my comment above with hints on how I usually make things fast and see if there is something worth trying.

**Comment by DanGrayson** [comment]
Here's something else I just noticed.  Suppose you use "unfold" on the type of something in the context, e.g., so you can what the components are so you know which projection to type.  That can slow down the "Defined" step at the end a lot.  So, although it might be tempting to leave the unfold statements in the proof for the convenience of the future reader, it is best to remove them after taking a peek.

**Comment by mortberg** [comment]
That should be related to what I say about not keeping simpl/cbn. Those tactics tend to unfold things a lot and the more things are unfolded the more Coq seems to compare things. So removing as many cbn, simpl and unfold as possible from the script once the proof is finished seems to make the Qed/Defined faster.

**Comment by DanGrayson** [comment]
The homological algebra files mostly have "Unset Kernel Term Sharing" in them.  Tomi, does that make them faster?  If so, that's also an indication that something is unfolding things deeply, and it might be a good idea to defeat the initial unfolding.  I'm not implying that I know how to do that or that I've ever been successful at it in a principled way.

**Comment by DanGrayson** [comment]
The slowest Qed in KA.v is that of ComplexHomotComm4, which is never used: 53 seconds.

**Comment by tpannila** [comment]
@DanGrayson , If you comment `Unset Kernel Term Sharing` in TranslationFunctors.v , then 
coq hangs on line 106. Here is a similar example using integers. Comment `Unset Kernel 
Term Sharing` so that coq  hangs on Lemma hangs.

```
Section test_bug.

  Definition test : UU := ∑ f : hz -> UU, ∏ (i : hz), (f i) -> (f (i + 1)).
  Definition mk_test (f : hz -> UU) (H : ∏ (i : hz), (f i) -> (f (i + 1))) : test :=
    tpair _ f H.
  Variable X : test.
  Definition translate_X : test :=
    mk_test (fun (i : hz) => (pr1 X) (i + 1)) (fun (i : hz) => (pr2 X) (i + 1)).
  
  Unset Kernel Term Sharing.
  Lemma hangs (i : hz) : (pr2 translate_X) i = (pr2 X) (i + 1).
  Proof. apply idpath. Qed.
  Set Kernel Term Sharing.

End test_bug.
```

I think Coq developers already know this. See
https://coq.inria.fr/bugs/show_bug.cgi?id=3309 Comment 3.

**Comment by mortberg** [comment]
@DanGrayson : Now that https://github.com/UniMath/UniMath/pull/652 has been merged, could you please make a new list of the 10 (or 20) slowest files? I remember seeing that some files in CT/limits got a lot slower and I want to investigate it, hopefully my findings can be used to speed up some of the other slow files.

**Comment by mortberg** [comment]
Maybe you can also share your timing script with us? I could write my own, but it seems easier if we use the same.

**Comment by DanGrayson** [comment]
You may wish to vary the -j parameter, depending on your CPU.  I've just been experimenting, and -j4 is even better than -j3 on mine (for elapsed real time).

The script:
```
set -x ; git log -1 && (cd sub/coq && git log -1) && make clean TAGS check-first && time make -j4 TIMECMD='gtime -f "user time %U: $*"' all 2>&1 | grep 'user time' | sort -r -k 3 -g
```

The top 10:
```
user time 200.71: UniMath/HomologicalAlgebra/TranslationFunctors
user time 171.74: UniMath/CategoryTheory/limits/graphs/pushouts
user time 137.80: UniMath/CategoryTheory/limits/graphs/pullbacks
user time 126.81: UniMath/HomologicalAlgebra/MappingCone
user time 116.65: UniMath/HomologicalAlgebra/KA
user time 113.56: UniMath/HomologicalAlgebra/KAPreTriangulated
user time 95.56: UniMath/CategoryTheory/LocalizingClass
user time 91.19: UniMath/HomologicalAlgebra/CohomologyComplex
user time 81.52: UniMath/HomologicalAlgebra/KATriangulated
user time 66.76: UniMath/CategoryTheory/limits/graphs/equalizers
```

**Comment by mortberg** [comment]
Thanks a lot! I took a quick look and the slowness in pushouts seem to come from some calls to `cbn` being very slow. I think they can be simply removed, I'll look into it tomorrow and make a PR.

**Comment by DanGrayson** [comment]
Some times the goal or the type of a hypothesis needs to be modified to get fragile tactics, such as "rewrite", to work.  One can use "cbn" privately to get the answer, and then write "change" instead to quickly get what you want in the code.

**Comment by DanGrayson** [comment]
By the way, Anders, the file INSTALL.md tells how to get the slowest proof steps from a file.

**Comment by mortberg** [comment]
I just made a pass removing slow calls to `cbn` in the file on pushouts that got very slow after merging the PR yesterday. The timing before my pass was:
```
user time 200.03: UniMath/CategoryTheory/limits/graphs/pushouts
```

and afterwards:

```
user time 1.32: UniMath/CategoryTheory/limits/graphs/pushouts
```

So by just removing slow calls to `cbn` I got the compilation time down to almost nothing from more than 3 minutes. 

I'll do the same change to some other slow files in CT and then make a PR.

**Comment by DanGrayson** [comment]
Great!

**Comment by JasonGross** [comment]
This sounds like a bug in cbn, if it was fast before primitive projections and slow after them.  Would it be possible to make a test case and report it?

**Comment by mortberg** [comment]
Good idea Jason, I'll try to make a simple self contained file illustrating this.

**Comment by DanGrayson** [comment]
Indeed, it was much quicker before.

**Comment by benediktahrens** [comment]
Somebody should summarize this discussion on a wiki page, with a reference to this issue. Afterwards, this issue should be closed. A new issue can be opened when necessary.

**Comment by benediktahrens** [comment]
I am closing this issue - new statistics would of course be very welcome!
